src/cmd/chat_mode.rs

```diff
@@ -0,0 +1,215 @@
+    msg: &Message,
+    response: &str,
+  ) -> Result<(), String> {
+    const MAX_MESSAGE_LENGTH: usize = 1000;
+
+    if response.len() <= MAX_MESSAGE_LENGTH {
+      msg
+        .reply(&ctx.http, response)
+        .await
+        .map_err(|e| format!("Failed to send reply: {}", e))?;
+    } else {
+      let (emote_name, bot_nick) = {
+        let config = Config::global_instance().read().unwrap();
+        (
+          config.emote_name.clone(),
+          ctx.cache.current_user().name.clone(),
+        )
+      };
+      let snarky_response = format!(
+        "{} doesn't really feel like speaking that much {}",
+        bot_nick, emote_name
+      );
+      msg
+        .reply(&ctx.http, &snarky_response)
+        .await
+        .map_err(|e| format!("Failed to send snarky response: {}", e))?;
+    }
+
+    Ok(())
+  }
+}
+
+#[async_trait]
+impl MessageListener for ChatModeHandler {
+  #[instrument(name = "ChatMode", level = "INFO", skip(self, ctx, msg))]
+  async fn message(&self, ctx: &Context, msg: &Message) {
+    // Skip if message is from a bot (including ourselves)
+    if msg.author.bot {
+      return;
+    }
+
+    // Only process if bot is mentioned
+    if !msg.mentions_me(ctx).await.ok().unwrap_or_default() {
+      return;
+    }
+
+    // Check if chat mode is enabled
+    let (chat_mode_enabled, emote_name, max_messages) = {
+      let config = Config::global_instance().read().unwrap();
+      (
+        config.chat_mode_enabled,
+        config.emote_name.clone(),
+        config.chat_mode_max_messages_per_conversation,
+      )
+    };
+
+    if !chat_mode_enabled {
+      let snarky_response = format!(
+        "Chat mode is currently disabled. {} Maybe try again later when I'm feeling more chatty.",
+        emote_name
+      );
+      if let Err(e) = msg.reply(&ctx.http, &snarky_response).await {
+        error!("Failed to send disabled chat mode response: {}", e);
+      }
+      return;
+    }
+
+    info!(
+      "Processing chat mode message from user: {}",
+      msg.author.name
+    );
+
+    // Build conversation chain using functional style
+    let conversation_messages: Vec<Message> = std::iter::once(msg)
+      .flat_map(|m| m.referenced_message.iter())
+      .map(|m| (**m).clone())
+      .take(max_messages)
+      .collect::<Vec<_>>()
+      .into_iter()
+      .rev()
+      .collect();
+
+    // Check if we've hit the max conversation limit
+    if conversation_messages.len() >= max_messages {
+      let response = "Max chat length reached, start a new message chain";
+      if let Err(e) = msg.reply(&ctx.http, response).await {
+        error!("Failed to send max length response: {}", e);
+      }
+      return;
+    }
+
+    // Get conversation key using functional style
+    let conversation_key: ConversationId = std::iter::once(msg)
+      .flat_map(|m| m.referenced_message.iter())
+      .map(|m| m.id.to_string())
+      .next_back()
+      .unwrap_or_else(|| msg.id.to_string());
+
```

Comment:
Use the first element of  `conversation_messages` for this, since that should be the original message in the chain at this point in the code. This removes an additional iteration.